**Importing Libraries**

In [1]:
import pandas as pd
import numpy as np

**Reading The File**

In [2]:
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Matiks.csv')

In [3]:
df.head()

User_ID          Username  \
0  7280e6c4-6f7c-45dd-a8fc-c58389ea8e07  geoffreyanderson   
1  23c48d4f-f5d0-4ff4-ba0f-2007441b9b57   riverachristian   
2  cf8d530c-c137-4346-a78b-e76e36d45e2a        brownchris   
3  47fcbe87-a1c1-40c3-b450-1b5692f61538     christopher90   
4  0b620a32-9e77-4b4a-9931-f0b654bef095          vfreeman   

                      Email Signup_Date   Country  Age  Gender Device_Type  \
0   haleymitchell@gmail.com   15-Dec-24   Austria   22   Other      Mobile   
1  masonmelissa@hotmail.com   07-Mar-24     Gabon   22   Other          PC   
2     mnichols@mcmillan.net   19-Oct-23   Ireland   36  Female          PC   
3         ttaylor@gmail.com   28-Sep-23   Belarus   23   Other          PC   
4        amanda80@gmail.com   09-Aug-24  Slovenia   26   Other          PC   

  Game_Title  Total_Play_Sessions  Avg_Session_Duration_Min  \
0  MysticWar                   21                      5.78   
1  QuestRaid                   22                     32.53   
2  QuestRaid                   12                     21.55   
3  QuestRaid                   19                     28.36   
4  QuestRaid                   18                     17.63   

   Total_Hours_Played  In_Game_Purchases_Count  Total_Revenue_USD Last_Login  \
0               38.41                        6              27.50  19-May-25   
1              129.02                       10               2.50  12-May-25   
2               62.50                        5              25.56  03-May-25   
3               79.40                        4             132.25  08-May-25   
4               86.70                        6              20.79  24-Apr-25   

  Subscription_Tier Referral_Source Preferred_Game_Mode Rank_Tier  \
0            Silver    Social Media               Co-op    Bronze   
1              Gold             Ads                Solo    Bronze   
2            Silver         Organic                Solo  Platinum   
3            Silver          Friend         Multiplayer   Diamond   
4              Free          Friend         Multiplayer   Diamond   

   Achievement_Score  
0               8721  
1               3572  
2               1060  
3               3591  
4               2666

**Checking  Variable Types**

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   User_ID                   10000 non-null  object 
 1   Username                  10000 non-null  object 
 2   Email                     10000 non-null  object 
 3   Signup_Date               10000 non-null  object 
 4   Country                   10000 non-null  object 
 5   Age                       10000 non-null  int64  
 6   Gender                    10000 non-null  object 
 7   Device_Type               10000 non-null  object 
 8   Game_Title                10000 non-null  object 
 9   Total_Play_Sessions       10000 non-null  int64  
 10  Avg_Session_Duration_Min  10000 non-null  float64
 11  Total_Hours_Played        10000 non-null  float64
 12  In_Game_Purchases_Count   10000 non-null  int64  
 13  Total_Revenue_USD         10000 non-null  float64
 14  Last_Lo

**Converting Sign Up Date and Login Date to Date Time**

In [5]:
for col in ['Last_Login', 'Signup_Date']:
    df[col] = pd.to_datetime(df[col],format="%d-%b-%y")

In [6]:
df.head()

User_ID          Username  \
0  7280e6c4-6f7c-45dd-a8fc-c58389ea8e07  geoffreyanderson   
1  23c48d4f-f5d0-4ff4-ba0f-2007441b9b57   riverachristian   
2  cf8d530c-c137-4346-a78b-e76e36d45e2a        brownchris   
3  47fcbe87-a1c1-40c3-b450-1b5692f61538     christopher90   
4  0b620a32-9e77-4b4a-9931-f0b654bef095          vfreeman   

                      Email Signup_Date   Country  Age  Gender Device_Type  \
0   haleymitchell@gmail.com  2024-12-15   Austria   22   Other      Mobile   
1  masonmelissa@hotmail.com  2024-03-07     Gabon   22   Other          PC   
2     mnichols@mcmillan.net  2023-10-19   Ireland   36  Female          PC   
3         ttaylor@gmail.com  2023-09-28   Belarus   23   Other          PC   
4        amanda80@gmail.com  2024-08-09  Slovenia   26   Other          PC   

  Game_Title  Total_Play_Sessions  Avg_Session_Duration_Min  \
0  MysticWar                   21                      5.78   
1  QuestRaid                   22                     32.53   
2  QuestRaid                   12                     21.55   
3  QuestRaid                   19                     28.36   
4  QuestRaid                   18                     17.63   

   Total_Hours_Played  In_Game_Purchases_Count  Total_Revenue_USD Last_Login  \
0               38.41                        6              27.50 2025-05-19   
1              129.02                       10               2.50 2025-05-12   
2               62.50                        5              25.56 2025-05-03   
3               79.40                        4             132.25 2025-05-08   
4               86.70                        6              20.79 2025-04-24   

  Subscription_Tier Referral_Source Preferred_Game_Mode Rank_Tier  \
0            Silver    Social Media               Co-op    Bronze   
1              Gold             Ads                Solo    Bronze   
2            Silver         Organic                Solo  Platinum   
3            Silver          Friend         Multiplayer   Diamond   
4              Free          Friend         Multiplayer   Diamond   

   Achievement_Score  
0               8721  
1               3572  
2               1060  
3               3591  
4               2666

**Summary Statistics**

In [7]:
df.describe()

Signup_Date           Age  Total_Play_Sessions  \
count                       10000  10000.000000         10000.000000   
mean   2024-05-22 18:18:43.200000     31.063700            19.993900   
min           2023-05-22 00:00:00     13.000000             6.000000   
25%           2023-11-24 00:00:00     22.000000            17.000000   
50%           2024-05-23 00:00:00     31.000000            20.000000   
75%           2024-11-22 00:00:00     40.000000            23.000000   
max           2025-05-21 00:00:00     49.000000            42.000000   
std                           NaN     10.687547             4.492314   

       Avg_Session_Duration_Min  Total_Hours_Played  In_Game_Purchases_Count  \
count              10000.000000        10000.000000              10000.00000   
mean                  30.035238          100.216981                  4.97600   
min                   -5.640000          -89.730000                  0.00000   
25%                   23.210000           66.490000                  3.00000   
50%                   29.860000          100.435000                  5.00000   
75%                   36.902500          133.900000                  6.00000   
max                   75.620000          283.260000                 17.00000   
std                   10.062647           49.642141                  2.23623   

       Total_Revenue_USD                     Last_Login  Achievement_Score  
count       10000.000000                          10000       10000.000000  
mean           50.177466  2025-05-06 09:51:59.040000256        5009.304400  
min             0.000000            2025-04-22 00:00:00           0.000000  
25%            14.900000            2025-04-29 00:00:00        2481.750000  
50%            34.760000            2025-05-06 00:00:00        5067.000000  
75%            69.225000            2025-05-14 00:00:00        7498.250000  
max           428.590000            2025-05-21 00:00:00        9997.000000  
std            49.600884                            NaN        2895.010702

There were negative values noticed in Avg_Session_Duration and Total_Hours_Played. Assuming these are typos as negative values for these variables do not make sense. Hence all negative values were converted to theor respective postiive values

In [8]:
df['Total_Hours_Played']=df['Total_Hours_Played'].apply(lambda x:abs(x))
df['Avg_Session_Duration_Min']=df['Avg_Session_Duration_Min'].apply(lambda x:abs(x))

Checking the summary statistics again to see how the changes made affected the distribution of the variables. Since there has not been a significant change in the mean value of the variables that were changes, i.e Total hours played and avg session duration it is safe to assume that the changes has not significantly affected the distribution of the variables

In [9]:
df.describe()

Signup_Date           Age  Total_Play_Sessions  \
count                       10000  10000.000000         10000.000000   
mean   2024-05-22 18:18:43.200000     31.063700            19.993900   
min           2023-05-22 00:00:00     13.000000             6.000000   
25%           2023-11-24 00:00:00     22.000000            17.000000   
50%           2024-05-23 00:00:00     31.000000            20.000000   
75%           2024-11-22 00:00:00     40.000000            23.000000   
max           2025-05-21 00:00:00     49.000000            42.000000   
std                           NaN     10.687547             4.492314   

       Avg_Session_Duration_Min  Total_Hours_Played  In_Game_Purchases_Count  \
count              10000.000000        10000.000000              10000.00000   
mean                  30.039052          100.944827                  4.97600   
min                    0.160000            0.000000                  0.00000   
25%                   23.210000           66.595000                  3.00000   
50%                   29.860000          100.435000                  5.00000   
75%                   36.902500          133.900000                  6.00000   
max                   75.620000          283.260000                 17.00000   
std                   10.051255           48.144706                  2.23623   

       Total_Revenue_USD                     Last_Login  Achievement_Score  
count       10000.000000                          10000       10000.000000  
mean           50.177466  2025-05-06 09:51:59.040000256        5009.304400  
min             0.000000            2025-04-22 00:00:00           0.000000  
25%            14.900000            2025-04-29 00:00:00        2481.750000  
50%            34.760000            2025-05-06 00:00:00        5067.000000  
75%            69.225000            2025-05-14 00:00:00        7498.250000  
max           428.590000            2025-05-21 00:00:00        9997.000000  
std            49.600884                            NaN        2895.010702

Created a new variable to assess difference between weekday and weekend traffic

In [10]:
df['Login_Weekend_Or_Weekday']=df['Last_Login'].dt.weekday.apply(lambda x:"Weekend" if x>=5 else "Weekday")

In [11]:
df.head()

User_ID          Username  \
0  7280e6c4-6f7c-45dd-a8fc-c58389ea8e07  geoffreyanderson   
1  23c48d4f-f5d0-4ff4-ba0f-2007441b9b57   riverachristian   
2  cf8d530c-c137-4346-a78b-e76e36d45e2a        brownchris   
3  47fcbe87-a1c1-40c3-b450-1b5692f61538     christopher90   
4  0b620a32-9e77-4b4a-9931-f0b654bef095          vfreeman   

                      Email Signup_Date   Country  Age  Gender Device_Type  \
0   haleymitchell@gmail.com  2024-12-15   Austria   22   Other      Mobile   
1  masonmelissa@hotmail.com  2024-03-07     Gabon   22   Other          PC   
2     mnichols@mcmillan.net  2023-10-19   Ireland   36  Female          PC   
3         ttaylor@gmail.com  2023-09-28   Belarus   23   Other          PC   
4        amanda80@gmail.com  2024-08-09  Slovenia   26   Other          PC   

  Game_Title  Total_Play_Sessions  ...  Total_Hours_Played  \
0  MysticWar                   21  ...               38.41   
1  QuestRaid                   22  ...              129.02   
2  QuestRaid                   12  ...               62.50   
3  QuestRaid                   19  ...               79.40   
4  QuestRaid                   18  ...               86.70   

   In_Game_Purchases_Count  Total_Revenue_USD  Last_Login Subscription_Tier  \
0                        6              27.50  2025-05-19            Silver   
1                       10               2.50  2025-05-12              Gold   
2                        5              25.56  2025-05-03            Silver   
3                        4             132.25  2025-05-08            Silver   
4                        6              20.79  2025-04-24              Free   

  Referral_Source Preferred_Game_Mode Rank_Tier Achievement_Score  \
0    Social Media               Co-op    Bronze              8721   
1             Ads                Solo    Bronze              3572   
2         Organic                Solo  Platinum              1060   
3          Friend         Multiplayer   Diamond              3591   
4          Friend         Multiplayer   Diamond              2666   

   Login_Weekend_Or_Weekday  
0                   Weekday  
1                   Weekday  
2                   Weekend  
3                   Weekday  
4                   Weekday  

[5 rows x 21 columns]

**Bucketing Age Variable To Study Age Group Wise User Behaviour**


In [31]:
bins = [12, 20, 30, 40, 50]
labels = ['13–20', '21–30', '31–40', '41–50']
df['Age_Groups'] = pd.cut(df['Age'], bins=bins, labels=labels)


**Bucketing Avg Session Duration in Minutes To Study Age Group Wise User Behaviour**

In [15]:
bins = [0,25,50,76]
labels_eng = ['Low Eng', 'Medium Eng', 'High Eng']
df['Engagement_Level'] = pd.cut(df['Avg_Session_Duration_Min'], bins=bins, labels=labels_eng)

Calculating Days Since Last Login- Can be USed to Understand the last time the user used the app. Can be used to understand user behaviour and segment them. Can also be used to understand if a user has churned or not.

Here the reference date for studying last login is taken as 21st  May because the most highest value of last login date in 21st May. So assuming here that the the data is available to this date only

In [32]:
date='21-05-2025'
date=pd.to_datetime(date)
df['Days Since Last Login']=(date-df['Last_Login']).dt.days

/tmp/ipython-input-32-306064896.py:2: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  date=pd.to_datetime(date)


In [33]:
df.head()

User_ID          Username  \
0  7280e6c4-6f7c-45dd-a8fc-c58389ea8e07  geoffreyanderson   
1  23c48d4f-f5d0-4ff4-ba0f-2007441b9b57   riverachristian   
2  cf8d530c-c137-4346-a78b-e76e36d45e2a        brownchris   
3  47fcbe87-a1c1-40c3-b450-1b5692f61538     christopher90   
4  0b620a32-9e77-4b4a-9931-f0b654bef095          vfreeman   

                      Email Signup_Date   Country  Age  Gender Device_Type  \
0   haleymitchell@gmail.com  2024-12-15   Austria   22   Other      Mobile   
1  masonmelissa@hotmail.com  2024-03-07     Gabon   22   Other          PC   
2     mnichols@mcmillan.net  2023-10-19   Ireland   36  Female          PC   
3         ttaylor@gmail.com  2023-09-28   Belarus   23   Other          PC   
4        amanda80@gmail.com  2024-08-09  Slovenia   26   Other          PC   

  Game_Title  Total_Play_Sessions  ...  Rank_Tier  Achievement_Score  \
0  MysticWar                   21  ...     Bronze               8721   
1  QuestRaid                   22  ...     Bronze               3572   
2  QuestRaid                   12  ...   Platinum               1060   
3  QuestRaid                   19  ...    Diamond               3591   
4  QuestRaid                   18  ...    Diamond               2666   

   Login_Weekend_Or_Weekday  Age_Groups Engagement_Level  \
0                   Weekday       21–30          Low Eng   
1                   Weekday       21–30       Medium Eng   
2                   Weekend       31–40          Low Eng   
3                   Weekday       21–30       Medium Eng   
4                   Weekday       21–30          Low Eng   

  Days Since Last Login Signed_Up Low Engagement Medium Engagement  \
0                     2         1              1                 1   
1                     9         1              1                 1   
2                    18         1              0                 0   
3                    13         1              1                 0   
4                    27         1              1                 0   

   High Engagement  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 28 columns]

In [34]:
df.describe()

Signup_Date           Age  Total_Play_Sessions  \
count                       10000  10000.000000         10000.000000   
mean   2024-05-22 18:18:43.200000     31.063700            19.993900   
min           2023-05-22 00:00:00     13.000000             6.000000   
25%           2023-11-24 00:00:00     22.000000            17.000000   
50%           2024-05-23 00:00:00     31.000000            20.000000   
75%           2024-11-22 00:00:00     40.000000            23.000000   
max           2025-05-21 00:00:00     49.000000            42.000000   
std                           NaN     10.687547             4.492314   

       Avg_Session_Duration_Min  Total_Hours_Played  In_Game_Purchases_Count  \
count              10000.000000        10000.000000              10000.00000   
mean                  30.039052          100.944827                  4.97600   
min                    0.160000            0.000000                  0.00000   
25%                   23.210000           66.595000                  3.00000   
50%                   29.860000          100.435000                  5.00000   
75%                   36.902500          133.900000                  6.00000   
max                   75.620000          283.260000                 17.00000   
std                   10.051255           48.144706                  2.23623   

       Total_Revenue_USD                     Last_Login  Achievement_Score  \
count       10000.000000                          10000       10000.000000   
mean           50.177466  2025-05-06 09:51:59.040000256        5009.304400   
min             0.000000            2025-04-22 00:00:00           0.000000   
25%            14.900000            2025-04-29 00:00:00        2481.750000   
50%            34.760000            2025-05-06 00:00:00        5067.000000   
75%            69.225000            2025-05-14 00:00:00        7498.250000   
max           428.590000            2025-05-21 00:00:00        9997.000000   
std            49.600884                            NaN        2895.010702   

       Days Since Last Login  Signed_Up  Low Engagement  Medium Engagement  \
count           10000.000000    10000.0    10000.000000       10000.000000   
mean               14.588900        1.0        0.777600           0.526300   
min                 0.000000        1.0        0.000000           0.000000   
25%                 7.000000        1.0        1.000000           0.000000   
50%                15.000000        1.0        1.000000           1.000000   
75%                22.000000        1.0        1.000000           1.000000   
max                29.000000        1.0        1.000000           1.000000   
std                 8.653183        0.0        0.415879           0.499333   

       High Engagement  
count     10000.000000  
mean          0.281600  
min           0.000000  
25%           0.000000  
50%           0.000000  
75%           1.000000  
max           1.000000  
std           0.449802

Classifying users into signed up, low engagement, medium engagement and high engagemment users. Can be used for understanding the user funnel as a user progresses from being signup to being a highly engaged user. Can be used to study where exaclty drop offs are happening and the corrective actions can be used to minimise these drop offs

In [19]:
df['Signed_Up'] = 1  # All signed up
df['Low Engagement'] = (df['Total_Play_Sessions'] >= 17).astype(int)
df['Medium Engagement'] = (df['Total_Play_Sessions'] >= 20).astype(int)
df['High Engagement'] = (df['Total_Play_Sessions'] >=23).astype(int)

In [35]:
import plotly.graph_objects as go

# Aggregating counts for each stage
signed_up = df['Signed_Up'].sum()
low = df['Low Engagement'].sum()
medium = df['Medium Engagement'].sum()
high = df['High Engagement'].sum()

# Define stages and values
stages = ["Signed Up", "Low Engagement", "Medium Engagement", "High Engagement"]
values = [signed_up, low, medium, high]

# Create the funnel chart
fig = go.Figure(go.Funnel(
    y = stages,
    x = values,
    textinfo = "value+percent previous+percent initial"
))

fig.update_layout(title="User Engagement Funnel – Matiks")
fig.show()


In [37]:
df.head()

User_ID          Username  \
0  7280e6c4-6f7c-45dd-a8fc-c58389ea8e07  geoffreyanderson   
1  23c48d4f-f5d0-4ff4-ba0f-2007441b9b57   riverachristian   
2  cf8d530c-c137-4346-a78b-e76e36d45e2a        brownchris   
3  47fcbe87-a1c1-40c3-b450-1b5692f61538     christopher90   
4  0b620a32-9e77-4b4a-9931-f0b654bef095          vfreeman   

                      Email Signup_Date   Country  Age  Gender Device_Type  \
0   haleymitchell@gmail.com  2024-12-15   Austria   22   Other      Mobile   
1  masonmelissa@hotmail.com  2024-03-07     Gabon   22   Other          PC   
2     mnichols@mcmillan.net  2023-10-19   Ireland   36  Female          PC   
3         ttaylor@gmail.com  2023-09-28   Belarus   23   Other          PC   
4        amanda80@gmail.com  2024-08-09  Slovenia   26   Other          PC   

  Game_Title  Total_Play_Sessions  ...  Rank_Tier  Achievement_Score  \
0  MysticWar                   21  ...     Bronze               8721   
1  QuestRaid                   22  ...     Bronze               3572   
2  QuestRaid                   12  ...   Platinum               1060   
3  QuestRaid                   19  ...    Diamond               3591   
4  QuestRaid                   18  ...    Diamond               2666   

   Login_Weekend_Or_Weekday  Age_Groups Engagement_Level  \
0                   Weekday       21–30          Low Eng   
1                   Weekday       21–30       Medium Eng   
2                   Weekend       31–40          Low Eng   
3                   Weekday       21–30       Medium Eng   
4                   Weekday       21–30          Low Eng   

  Days Since Last Login Signed_Up Low Engagement Medium Engagement  \
0                     2         1              1                 1   
1                     9         1              1                 1   
2                    18         1              0                 0   
3                    13         1              1                 0   
4                    27         1              1                 0   

   High Engagement  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 28 columns]

**RFM Study Of Users**

A RFM segmentation of users has been done in the further steps.

R-Recency the last time a user has used the app. A larger value is bad and typically indicates the customer is moving towards churn.

F-Frequency: The number of times the user has used the app. Here total play sessions has been used to as to model the frequency

M- Monetary Value: The amount spent by a user on various in app purchases and/or for signing in on the app. Can be used to identify the high spenders and low spenders as both these users need different strategies to ensure retention

The three variables metioned above were cut into 3 based on quatile based cuts. Based on these quantile cuts  scores have been assigned

In [38]:
df_recency = pd.qcut(df['Days Since Last Login'],q=3)
df_recency.value_counts()

Days Since Last Login
(-0.001, 10.0]    3607
(10.0, 20.0]      3365
(20.0, 29.0]      3028
Name: count, dtype: int64

As from above

Since a lower days since last login is better

Users with days since last login from 0 to 10 wil be given a Recency score of 3

Users with days since last login from 11 to 20 wil be given a Recency score of 2

Users with days since last login from 21 to 30 wil be given a Recency score of 1



In [40]:
df_frequency = pd.qcut(df['Total_Play_Sessions'],q=3)
df_frequency.value_counts()

Total_Play_Sessions
(5.999, 18.0]    3874
(18.0, 22.0]     3310
(22.0, 42.0]     2816
Name: count, dtype: int64

In [39]:
df_monval = pd.qcut(df['Total_Revenue_USD'],q=3)
df_monval.value_counts()

Total_Revenue_USD
(54.94, 428.59]    3334
(-0.001, 20.74]    3333
(20.74, 54.94]     3333
Name: count, dtype: int64

Since a higher amount spent  is better

Users with Total Revenue Per User from 0 to 20$ wil be given a Monetary Value score of 1

Users with Total Revenue Per User from 21 to 54$ wil be given a Monetary Value score of 2

Users with Total Revenue Per User from 55 to 429$ wil be given a Monetary Value score of 3



**Assigning RFM Scores To Users Based On Quantile Cuts**

Since there will 3x3x3=27 customer segments there will be 27 customer segments. This is too huge to study. Hence based on RFM scores, 4 segments were defined as below


In [24]:
# Copy the working DataFrame
rfm_df = df.copy()

#Define RFM metrics
rfm_df['Recency'] = df['Days Since Last Login']                 # Lower is better
rfm_df['Frequency'] = df['Total_Play_Sessions']                # Higher is better
rfm_df['Monetary'] = df['Total_Revenue_USD']                   # Higher is better

#Quantile-based binning (into 3 levels: Low=1, Mid=2, High=3)
# Reverse Recency bins (because lower recency = better)
rfm_df['R_Score'] = pd.qcut(rfm_df['Recency'], 3, labels=[3, 2, 1]).astype(int)
rfm_df['F_Score'] = pd.qcut(rfm_df['Frequency'], 3, labels=[1, 2, 3]).astype(int)
rfm_df['M_Score'] = pd.qcut(rfm_df['Monetary'], 3, labels=[1, 2, 3]).astype(int)


#Assign customer segment

def assign_segment(row):
    if row['R_Score'] == 3 and row['F_Score'] ==3 and row['M_Score'] == 3:
        return 'CBest Customers'
    elif row['M_Score'] == 3 and row['F_Score'] >= 2 and row['R_Score']>=2:
        return 'Big Spenders'
    elif row['F_Score'] == 1 and row['R_Score']==3:
        return 'Newbies'
    elif row['F_Score'] >= 2 and row['R_Score'] >= 2:
        return 'Loyal Customers'
    elif row['R_Score'] <= 2 and row['F_Score'] <= 2 :
        return 'Churned'
    #elif row['R_Score'] == 1 and row['F_Score'] <= 2 and row['M_Score'] <= 2:
     #   return 'Churned'
    else:
        return 'Others'  # fallback


rfm_df['Customer_Segment'] = rfm_df.apply(assign_segment, axis=1)

# Step 4: Final columns for export or Power BI use
rfm_result = rfm_df[[
    'User_ID', 'Recency', 'Frequency', 'Monetary',
    'R_Score', 'F_Score', 'M_Score',  'Customer_Segment'
]]

# Optional: Save to CSV for Power BI
# rfm_result.to_csv("RFM_Segments.csv", index=False)

# Preview
rfm_result.head()


User_ID  Recency  Frequency  Monetary  \
0  7280e6c4-6f7c-45dd-a8fc-c58389ea8e07       11         21     27.50   
1  23c48d4f-f5d0-4ff4-ba0f-2007441b9b57       18         22      2.50   
2  cf8d530c-c137-4346-a78b-e76e36d45e2a       27         12     25.56   
3  47fcbe87-a1c1-40c3-b450-1b5692f61538       22         19    132.25   
4  0b620a32-9e77-4b4a-9931-f0b654bef095       36         18     20.79   

   R_Score  F_Score  M_Score Customer_Segment  
0        3        2        2  Loyal Customers  
1        3        2        1  Loyal Customers  
2        2        1        2          Churned  
3        2        2        3     Big Spenders  
4        1        1        2          Churned

In [ ]:
#Assign customer segment

def assign_segment(row):
    #if row['R_Score'] == 3 and row['F_Score'] >= 2 and row['M_Score'] == 3:
        #return 'Champions'
    if row['M_Score'] == 3 and row['F_Score'] >= 2 and row['R_Score']>=2:
        return 'Big Spenders'
    elif row['F_Score'] == 1 and row['R_Score']==3:
        return 'Newbies'
    elif row['F_Score'] >= 2 and row['R_Score'] >= 2:
        return 'Loyal Customers'
    elif row['R_Score'] <= 2 and row['F_Score'] <= 2 :
        return 'Churned'
    #elif row['R_Score'] == 1 and row['F_Score'] <= 2 and row['M_Score'] <= 2:
     #   return 'Churned'
    else:
        return 'Others'  # fallback


rfm_df['Customer_Segment'] = rfm_df.apply(assign_segment, axis=1)

# Step 4: Final columns for export or Power BI use
rfm_result = rfm_df[[
    'User_ID', 'Recency', 'Frequency', 'Monetary',
    'R_Score', 'F_Score', 'M_Score',  'Customer_Segment'
]]

# Optional: Save to CSV for Power BI
# rfm_result.to_csv("RFM_Segments.csv", index=False)

# Preview
rfm_result.head()


In [25]:
new_df=pd.merge(df,rfm_result,on='User_ID')

In [26]:
new_df.groupby('Customer_Segment')['Days Since Last Login'].mean()

Customer_Segment
Big Spenders       18.804490
Churned            30.488278
Loyal Customers    19.164003
Newbies            14.046283
Others             34.184884
Name: Days Since Last Login, dtype: float64

In [27]:
new_df.groupby('Customer_Segment')['Total_Play_Sessions'].mean()

Customer_Segment
Big Spenders       22.790007
Churned            17.012156
Loyal Customers    22.760598
Newbies            15.622020
Others             25.473256
Name: Total_Play_Sessions, dtype: float64

In [28]:
new_df.groupby('Customer_Segment')['Total_Revenue_USD'].mean()

Customer_Segment
Big Spenders       105.095351
Churned             50.144365
Loyal Customers     22.458937
Newbies             53.250281
Others              49.787767
Name: Total_Revenue_USD, dtype: float64

In [29]:
df.to_csv("Matiks_New.csv", index=False)

In [30]:
rfm_result.to_csv("RFM_Segments.csv", index=False)